### DISCUSSION TITLE
Control z index of connection line

### DISCUSSION THREAD

**Discussion by RunDevelopment** [xyflow/xyflow#2158] [New Features]
Right now, it is not possible to control the z index of the connection line **independently of the z index of edges**. The connection line simply gets assigned to [the group with the highest z index](https://github.com/wbkd/react-flow/blob/aab08582124046318ab0552f0ce8bcbee62ea9be/src/container/EdgeRenderer/index.tsx#L182).

### Our use case

In our use case, all edges have a z index of 0 and all nodes have a z index of 1 (I'm simplifying a little). So edges are always below nodes. This ensures that the user can always interact with the form inputs in our nodes. However, we want the connection line to always be above all nodes, so the user can see what they are connecting.

## New feature

I can imagine 3 ways to solve this.

### 1: `connectionLineZIndex`

Add a new `connectionLineZIndex?: number` property to the `ReactFlow` component. This property is optional and will control the z index of the connection line. If the property is missing, everything will behave as it currently does. If the property is present, the connection line will in a new `<svg>` element with the given z index (or it may be assigned to an edge group with the given z index, if such a group exists).

This is a new feature and will only require a new minor version.

### 2: Use `connectionLineStyle.zIndex`

As explained below, the `zIndex` property currently does nothing. Admittedly, this is quite hacky, but we could read the `zIndex` (if present) create a new `<svg>` element with this z index (just like this `connectionLineZIndex`).

This is breaking change as the `zIndex` now affects the behavior of the program. It is no longer a no-op.

### 3: `connectionLineContainerStyle`

Add a new `connectionLineContainerStyle?: CSSProperties` property to the `ReactFlow` component. This property controls the styles of the `<svg>` container element of the connection line. If the property is not present, then everything will behave as it currently does. If it is present, then a new `<svg>` container will be created for the connection line and the styles will be applied.

This is a new feature and will only require a new minor version.

## Alternatives

In #810, it was suggested to raise the z index of all edges (in this case, via CSS). This does not work for us, because all edges should remain below all nodes. Only the connection line should be above all nodes.

`<ReactFlow connectionLineStyle={{ zIndex: 1000 }}` also does not work, because it controls the z index of connection line SVG elements (e.g. `<g>`, `<path`>, etc.) inside the SVG document (`<svg>`). Since [the `zIndex` property has no effect on the paint order](https://stackoverflow.com/questions/17786618/how-to-use-z-index-in-svg-elements), it does nothing.

Currently, there is no way to achieve the desired effect that I am aware of.

---

I prefer solution 3 as it seems to be the most generally useful. No matter solution we pick, **I am willing to make a PR implementing it**.

**Comment by doflo-dfa** [comment]
It may also be useful to provide an API to place additional items in the same "stack" at the same level as the current parent and thus same z-indexing context  

Our use case is that you have all of the connector lines at the same z-index but you may have labels that you would like to have a superior position to the other lines that are currently constrained by the parent element (in this case the edge elements group <g>) z-index. exploring a work around now but it feels like a hack

**Reply by RunDevelopment** [reply]
@doflo-dfa You might want to create a new discussion for this, but I already implemented my requested feature some time ago (#2159).
